Src/IronPython/Runtime/PythonContext.cs

```diff
@@ -936,125 +936,79 @@ private sealed class OptimizedBuiltinHasher {
             ContractUtils.Requires(stream.CanSeek && stream.CanRead, nameof(stream), "The stream must support seeking and reading");
 
             stream.Seek(0, SeekOrigin.Begin);
-            byte[] bomBuffer = new byte[3];
-            int bomRead = stream.Read(bomBuffer, 0, 3);
-            int bytesRead = 0;
-            bool isUtf8 = false;
-            if (bomRead == 3 && (bomBuffer[0] == 0xef && bomBuffer[1] == 0xbb && bomBuffer[2] == 0xbf)) {
-                isUtf8 = true;
-                bytesRead = 3;
-            } else {
-                stream.Seek(0, SeekOrigin.Begin);
-            }
 
+            Encoding sourceEncoding = null;
             string encodingName = null;
             int linesRead = 0;
+            bool hasBom = false;
             // sr is used to read the magic comments (PEP-263)
-            // default system ASCII encoding will never throw exceptions, converting bad bytes to '?' instead
-            // which is sufficient to parse the magic comments
-            // also StreamReader will automatically switch to proper Unicode encoding (non-throwing) when BOM present
-            using (StreamReader sr = new StreamReader(stream, Encoding.ASCII, detectEncodingFromByteOrderMarks: true, bufferSize: 1024, leaveOpen: true)) {
+            // Default system ASCII encoding will not throw exceptions, converting bad bytes to '?' instead
+            // which is sufficient to correctly parse the magic comments.
+            // Also StreamReader will automatically switch to proper Unicode encoding (non-throwing) when BOM present.
+            Encoding bootstrapEncoding = Encoding.ASCII;
+            using (StreamReader sr = new StreamReader(stream, bootstrapEncoding, detectEncodingFromByteOrderMarks: true, bufferSize: 1024, leaveOpen: true)) {
                 string line;
-                line = ReadOneLine(sr, ref bytesRead);
+                line = sr.ReadLine();
                 linesRead++;
 
                 // magic encoding must be on line 1 or 2
                 // if there is any encoding specified on line 1 (even an invalid one), line 2 is ignored
-                if (line != null && (encodingName = Tokenizer.GetEncodingNameFromComment(line)) == null) {
+                if ((encodingName = Tokenizer.GetEncodingNameFromComment(line)) == null) {
```

Comment:
Any specific reason for remove the `line != null`? I guess it's an insignificant optimization...

Reply:
IMHO simpler code and moves the check to `GetEncodingNameFromComment`, which becomes more robust to input.

Reply:
In CPython, the second line is not inspected if there is no `\n` or `\r` within the first 512 characters of the first line.
Perhaps this is an optimization to avoid scanning large binary files.
So the test `line != null` has to stay. I will add a test case to cover it.